### ISSUE TITLE
WHDLoad.hdf Question

### ISSUE THREAD

**Issue by MajorWilly** [libretro/libretro-uae#191]
I have a bunch of WHDLoad HDF's but the slave files are in a directory in each HDF. What do I need to add to the bottom of the Startup Sequence in the WHDLoad.hdf so that it searches for the slave inside a folder ?

This is what it currently looks like but is it only searching the root ?
`list #?.slave quick nohead to ENV:slave`

Example of the root of one of my HDFs -

```
AddamsFamily (directory. slave is inside here)
AddamsFamily.info
```

**Comment by sonninnos** [comment]
I've been needing to figure out how to do that for the perhaps arriving LHA support to ever work.

So using the same method this should do it:
`list #? quick dirs nohead to ENV:dirs` and then `cd $dirs` should just pick the first and most likely only one.

I'll include it in the next WHDLoad.hdf update, which is soon coming anyway.

**Comment by MajorWilly** [comment]
worked great, thanks.

I don't know which ones, but the WHDLoad page mentions that some games need XPK libraries `xpkmaster.library` because of packed files, so you might wan't to add that.

**Comment by MajorWilly** [comment]
One more thing if possible. A few slaves filenames are 1 character too long for Amiga FFS so the name gets cut to `gamename.slav`. Is there a way to also search for `.slav` ? (or make the last 'e' a wildcard ?)

**Comment by sonninnos** [comment]
Would you like to take this new hdf for a test drive to see if there is still something missing?
Thanks!

[WHDLoad.zip](https://github.com/libretro/libretro-uae/files/4012440/WHDLoad.zip)

**Comment by MajorWilly** [comment]
With the new WHDLoad.hdf, if I try to use it without copying Kickstarts to it, it says `Error validating WHDLoad` and then `Can't open DH0:Devs/Kickstarts/`. If I copy the Kickstarts to it then try a WHDLoad game HDF then it works. Slaves inside directories work and Slaves named `.slav` work.

If I try to run a non-WHDLoad HDF with the WHDLoad core option ON then I still get errors about missing files. Is it just trying to find all the required files in the WHDLoad.hdf ?

Example -
[ABD The Next Break AGA.zip](https://github.com/libretro/libretro-uae/files/4012592/ABD.The.Next.Break.AGA.zip)
(Works fine if the WHDLoad core option is OFF)

**Comment by sonninnos** [comment]
Did the automatic Kickstart copy work before with that platform (Android I assume)?

That issue should be fixed now: [WHDLoad.zip](https://github.com/libretro/libretro-uae/files/4012664/WHDLoad.zip)

**Comment by MajorWilly** [comment]
Several say this before loading but they run OK -
```
RAM:T already exists
RAM:ENV already exists
```
A few have said -
`Insert volume KEYMAPS in any drive`

Not having Kickstarts in WHDLoad.hdf works again on PC but still crashes on Android.

**Comment by sonninnos** [comment]
Maybe this time's the charm:
[WHDLoad.zip](https://github.com/libretro/libretro-uae/files/4012835/WHDLoad.zip)

Nothing I can do about that directory filesystem not working at all issue from this HDF angle (except not bugging with requesters when the volume is not there). That would need either fixing the directory filesystem for the other platforms or removing it in favor of ADFlib and do the Kickstart copying outside the UAE. I'm lurking at the latter option.

**Comment by MajorWilly** [comment]
Almost there !

Still get a keymaps error. It was looking for the Keymaps on the Game HDF (devs/Keymaps).

Android needing the Kickstarts on the WHDLoad.hdf is no big deal. Just mention it in the Readme that it needs copying manually (with ADF Opus or run a WHDLoad game once on PC then copy the WHDLoad.hdf to Android. Not sure what ADFlib is). I don't know if this is also related to the Switch not finding the System directory (I don't own one) #194

**Comment by sonninnos** [comment]
Please share the keymap error HDF.

Yes the System volume is of course related. The point of ADFlib would be to have proper access to HDFs so the file operations can be done by the core before starting the emulation and not in the emulator.

So with Android you did not get those requesters about the missing volume even before this new HDF? It just crashes, before or after Kickstart?

**Comment by MajorWilly** [comment]
Keymap error is still there. An example HDF (I don't think the game works on PUAE) -
[Amiga Resident Evil AGA.zip](https://github.com/libretro/libretro-uae/files/4012986/Amiga.Resident.Evil.AGA.zip)
Also you'll notice that there are two 'system' directories and a 'whdload' directory on the Workbench. Is that the Retroarch system directory !? They don't appear if WHDLoad core option is OFF.

And yes, on Android it crashes at the beginning of loading (when trying to copy the kickstarts I suppose).

**Comment by sonninnos** [comment]
Finally ready for the PR I hope:
[WHDLoad.zip](https://github.com/libretro/libretro-uae/files/4013126/WHDLoad.zip)

Of course the RA's System is only there with the WHDLoad.hdf option, because it won't get accessed otherwise.

Does Android crash with having the `WHDLoad.prefs` and/or `.key` in system?

**Comment by MajorWilly** [comment]
I manually added the `WHDLoad.prefs` file to the System directory (on Android). It still crashes.  I also set Config+Splash in the core options.

**Comment by sonninnos** [comment]
The point was to make it crash, not to not crash. The core option setting does not matter in this case, as the `.prefs` file will be copied and overwritten always if it exists in system.

Well that one is weird. The guide is launching with `multiview` which is in the path and works from the CLI..

**Comment by MajorWilly** [comment]
I think I misunderstood you. This time, (on Android) I copied the WHDLoad.hdf with Kickstarts inside and the WHDLoad.prefs file. It crashed. I deleted the prefs file and it works.

**Comment by sonninnos** [comment]
Yeah I just wanted to know if there was any difference if the copied file was tiny and plain text, but apparently not.

I found a fix for the `multiview` thing, but it can't be fixed in the HDF alone, so I shall declare it finished enough for the PR. The reason is that the image has the same "System" label as the filesystem mount, so changed it to "RASystem".

**Comment by MajorWilly** [comment]
There was something that i've never tried before and that is saving a game. I tried Monkey Island 2 and made an ingame save F5. When I closed and reloaded the game it said `Error validating WHDLoad - Block 4836 out of range`. If you click cancel you get a couple more errors about Validation but the game loaded and the save loaded. Is it corrupting the WHDLoad.hdf when saving to it ? I tried on PC & Android and the same thing happened.

**Comment by sonninnos** [comment]
I only managed to achieve that once on the new separate WHDSaves.hdf, but only once, so not much to go on. Could you try if raising the `WriteDelay` setting in `WHDLoad.prefs` makes any difference?

**Comment by MajorWilly** [comment]
I deleted the `WriteDelay=0` from the bottom of the WHDLoad.prefs. Saved a game (which took longer as expected). I quit then reloaded and didn't get any Validation errors.

**Comment by sonninnos** [comment]
Hopefully that is it and hopefully there is a golden value between that 0 and default 150 to be found.

https://youtu.be/14ycMqFysJ0

**Comment by sonninnos** [comment]
I did some testing with `WriteDelay` doing saves on The Settlers, and so far looks like value 50 will be the new default. Random save corruption with values below it, but none with it.

I've been using value 0 for a long time with the directory filesystem, and had no problems. And I've also used value 0 on the real Amiga for ages, but then again I have PFS3 there and  WriteCache enabled (no need to shutdown without graceful QuitKey), so maybe this is an issue only with FFS and OS switching with `NoWriteCache`.

It sure is fast at 0, but too bad it is too fast.

**Comment by MajorWilly** [comment]
Found a couple of non whdload games that say `Can't find DH1:C`
[Super Street Fighter II - The New Challengers AGA.zip](https://github.com/libretro/libretro-uae/files/4014252/Super.Street.Fighter.II.-.The.New.Challengers.AGA.zip)
I know there is a WHDLoad version of this poor port but I thought i'd point out the error.

A couple of games doesn't run after clicking -
[Prototype.zip](https://github.com/libretro/libretro-uae/files/4014254/Prototype.zip)

I had a game that couldn't load a font, and 2 that said Software Failure - Program failed.

They all work without the WHDLoad core option. This could be endless. You'll end up mounting everything!

**Comment by sonninnos** [comment]
That SSF2 does something a lot more sinister than just not having a C directory. It has DH0: hardcoded in Startup-Sequence, which will never point to the correct drive as long as WHDLoad.hdf is DH0:. Oh well, not a big deal to rename it, I knew this day would come.

The problem with the Prototype is not apparent right now.

**Comment by sonninnos** [comment]
I got that Prototype working with assigning fonts over too. Included in the next batch.

**Comment by MajorWilly** [comment]
This also makes the non-WHDLoad HDF's of 'Star Trek' and 'Klondike' work and gives the correct font on 'Megaball'. I think that's pretty much all of them now.

**Comment by MajorWilly** [comment]
As MultiView is included in the WHDLoad.hdf is it possible to have an option (in the custom options when quitting) to open it so we can view the instructions or readme's that may say what the custom options are etc. ? Is that possible or won't it be able to navigate to the read only LHA DH1: ?

**Comment by sonninnos** [comment]
A read only hard drive is just as navigatable as any. I thought why not just show the readme always before MkCustom, but then again it might not be enough since not all games allow quitting even with the QuitKey, for example that Dungeon Master.

**Comment by MajorWilly** [comment]
That's a shame. It would have been good to have a choice of what file to view as a lot also come with the instructions and other stuff.

**Comment by sonninnos** [comment]
I replaced it with MuchMore, runs always before MkCustom. Included in the next batch.

**Comment by MajorWilly** [comment]
Nice one.

**Comment by MajorWilly** [comment]
A couple more requirements !?

The WHDLoad game demo `ACSYSDemo_v1.11_AGA` gives this error -
![CpWz_004](https://user-images.githubusercontent.com/52862903/71974101-8f170880-3208-11ea-85d3-7337f2a4969a.png)

With MuchMore, if I press 'L' to load a new file (like the instructions etc.) it says that the `asl.library` is missing. I tried the one from Workbench 3.1 and that seems compatible with both A600 and A1200 core configs. Also, the text wraps to the next line. Is the font too big ?

**Comment by sonninnos** [comment]
Looks like a bad slave based on this: https://github.com/midwan/amiberry/issues/442

> AFAIK, the OSemu.400 is usually included with the game data...
> ... having checked the file, it's in the data folder. So like Jaguar XJ220 this slave doesnt set the data path correctly from it's slave header. You will need to add DATA=DATA to the WHDLoad commands, and i've just tested that this works.

I'll have a look at that asl.library, even though I don't really see the need. Sure there are often manuals and other text files, but this may not be the best place to read anything else except the available custom parameters.

**Comment by MajorWilly** [comment]
May as well add the `asl.library` (for shits and giggles :smile: ) Is there anything that can be done about the text wrapping ? Maybe reduce the font size, or is that not possible (and wasting your time!).

For ACSYSDemo I added `DATA=DATA` to the MkCustom and it works. Do you think these should be added to the WHDSaves.hdf to make them a permanent thing ? (As well as `Custom1=3` for BattleIsle&DataDisks which is needed (and possibly for regular BattleIsle)).

**Comment by sonninnos** [comment]
I don't like the idea of doing any game specific exceptions and fixes, but that data-dir thing appears to be global enough.

Here again the `DATA=data` is included in the tooltypes, so again there won't be a problem launching via the icon. That is why the `wbrun` solution when there is no `custom` would be best for the long run. The only problem with that is that I have to find a simple way to remove ".info" from the end of the filename, since it won't launch with it.

It sure is not easy to find a legible font at small sizes.

But in the meantime here are those previous issues sorted, I'll let you hunt some more before I bother the overseers again with a PR: 
[WHDLoad.zip](https://github.com/libretro/libretro-uae/files/4035740/WHDLoad.zip)

**Comment by MajorWilly** [comment]
What are your personal thoughts on this font ? It seems a bit fat & thin if you know what I mean ? Vertical lines seem very skinny.

Indeed, I found another issue. Not sure if it's related to the WHDLoad.hdf. The game `Cadaver&ThePayoff_v2.4_0900` crashes the core/Retroarch.

**Comment by sonninnos** [comment]
I'll restore the default font size, this current can't be read with lores.

Yep, the 0 byte file seems to be the problem. 7z shows it as a file, but WinUAE shows it as a directory (`saves` in Cadaver). I removed it, made a new LHA and it worked fine. So I'll do some comparing with what WinUAE does different with LHAs. So not related to WHDLoad, it crashes right at attaching the LHA.

**Comment by MajorWilly** [comment]
WinUAE uses a UAEFSDB file for dealing with illegal characters/files etc. Is that what PUAE uses ?

http://eab.abime.net/showthread.php?t=5126

**Comment by MajorWilly** [comment]
PPMore has a much better built-in font -

![CpWz_005](https://user-images.githubusercontent.com/52862903/71999720-d8338080-3239-11ea-9c48-0a8d799af6ee.png)

**Comment by sonninnos** [comment]
But isn't that the same regular Topaz..

**Comment by MajorWilly** [comment]
Not sure what the font is but it looks a lot better, fullscreen and is also readable in LowRes. Not sure what happened with MuchMore.

**Comment by sonninnos** [comment]
Hmm, MuchMore does use that same font with my A1200 setup:
![winuae_2020_01_08_19_44_05_044](https://user-images.githubusercontent.com/45124675/72002001-47679f80-324f-11ea-8174-dcf44d032f6b.png)

But yep something is clearly missing since it looks very chunky here:
![retroarch_2020_01_08_19_30_52_184](https://user-images.githubusercontent.com/45124675/72002017-4fbfda80-324f-11ea-85a1-3c51c6de6c59.png)

**Comment by MajorWilly** [comment]
Does MuchMore do much more than PPMore ? :smile: May as well just use PPMore.

**Comment by sonninnos** [comment]
Same chunky Topaz 9 default with PPMore..

**Comment by MajorWilly** [comment]
I used this one -
[PPMore.zip](https://github.com/libretro/libretro-uae/files/4036775/PPMore.zip)

I got it from one of the non-WHDLoad HDF's. Not sure what version it is. It doesn't look like the screenshot you posted.

**Comment by sonninnos** [comment]
Oh heh, that PPMore is actually MuchMorePoPa 2.7, and this current is MuchMore 4.6. Replaced.

**Comment by MajorWilly** [comment]
Cool. I'll open a different issue about the games with 0 byte files crashing.

**Comment by sonninnos** [comment]
The next incarnation has a file requester feature when fire button is held down at boot and WBRun. The flow is as follows:
- If there is no `custom` (the default state):
  - Joystick button held down
     - Present `.info` select requester
     - WBRun the selected
  - No button held down
    - Tries to find the appropriate `.info` based on the `.slave`
    - If found, WBRun it
    - If not found (BattleIsle etc. with multiple icons), present `.info` select requester

- If there is `custom`:
  - Joystick button held down
    - Show `ReadMe`
    - Run `MkCustom`
    - Run game
  - No button held down
    - Run game

Please try to make this break also: 
[WHDLoad.zip](https://github.com/libretro/libretro-uae/files/4037232/WHDLoad.zip)

**Comment by MajorWilly** [comment]
Wow ! Very very impressed. 

3D Pool (has a crappy Pool clip)
Battle Isle (Different missions)
Bart vs Space Mutants (Watch intro or straight to game)

I broke it !

Pinball Prelude (after choosing a table in the requester it says `Couldn't load "LHA:PinballPrelude/PinballPrelude"` `object not found`). If you type in MkCustom `custom1=0` (or 1 or 2) then it works.

Haven't tried anything else yet. Overall, I give it a 9.5 out of 10. :smile: Awesome.

**Comment by sonninnos** [comment]
Phew, that was simple. It was because of the space in the `.info` filename. 

Here we go: [WHDLoad.zip](https://github.com/libretro/libretro-uae/files/4037519/WHDLoad.zip)

**Comment by MajorWilly** [comment]
The game might be crap but *cough*`DeluxeStripPoker2&DataDisks_v1.0`*cough* gives this error -

![CpWz_006](https://user-images.githubusercontent.com/52862903/72020815-85bb8980-3264-11ea-85be-afc5a77de6be.png)

It mentions XFD package in the readme.

**Comment by MajorWilly** [comment]
Can't find anything else at the moment. Overall, very nice update, and love the dark theme !

**Comment by MajorWilly** [comment]
Last one, off to bed -

GianaSistersSE_v1.2

`Tool not found "IconX"`

**Comment by sonninnos** [comment]
DeluxeStripPoker seems to be emulation accuracy related. Works fine in WinUAE with the same HDFs.

IconX included. 
[WHDLoad.zip](https://github.com/libretro/libretro-uae/files/4037956/WHDLoad.zip)
